File: packages/browser-host/src/input-handoff.js
// Asynchronous input acquisition to committed-snapshot handoff
// (browser side). Acquisition runs asynchronously OUTSIDE any semantic
// callback; the acquired rows and probed-missing keys are handed to the
// WASM host in one synchronous commit; later synchronous observations
// answer from the committed rows only. A requested-but-unacquired key
// raises typed NeedInputs naming the canonical — never a synchronous
// fetch inside a resolver.

/** Typed demand for the next asynchronous acquisition wave. */
export class NeedInputsError extends Error {
  /** @param {string[]} keys canonicals the committed snapshot never acquired */
  constructor(keys) {
    super(`NeedInputs: ${[...keys].join(", ")}`);
    this.name = "NeedInputsError";
    this.keys = [...keys];
  }
}

/**
 * Validate one acquisition wave before it is committed. Returns the
 * list of failures; an empty list means the wave is committable.
 * Fail-closed: a duplicate canonical, an overlap between acquired
 * and probed-missing, or a non-string row is reported, never silently
 * collapsed.
 *
 * @param {Array<{canonical: string, content: string}>} files
 * @param {string[]} missing
 * @returns {string[]}
 */
export function acquisitionWaveFailures(files, missing) {
  const failures = [];
  if (!Array.isArray(files)) return ["files is not an array"];
  if (!Array.isArray(missing)) return ["missing is not an array"];

  const seen = new Map();
  for (const [index, row] of files.entries()) {
    if (row == null || typeof row !== "object") {
      failures.push(`files[${index}] is not a row object`);
      continue;
    }
    if (typeof row.canonical !== "string" || row.canonical.length === 0) {
      failures.push(`files[${index}].canonical is not a non-empty string`);
      continue;
    }
    if (typeof row.content !== "string") {
      failures.push(`files[${index}].content is not a string`);
      continue;
    }
    const prior = seen.get(row.canonical);
    if (prior === "file") {
      // Identical duplicates collapse host-side; distinct contents are
      // refused host-side. Either way the wave stays well-formed here.
      continue;
    }
    if (prior === "missing") {
      failures.push(`canonical '${row.canonical}' is both acquired and probed missing`);
      continue;
    }
    seen.set(row.canonical, "file");
  }
  for (const [index, canonical] of missing.entries()) {
    if (typeof canonical !== "string" || canonical.length === 0) {
      failures.push(`missing[${index}] is not a non-empty string`);
      continue;
    }
    if (seen.get(canonical) === "file") {
      failures.push(`canonical '${canonical}' is both acquired and probed missing`);
      continue;
    }
    seen.set(canonical, "missing");
  }
  return failures;
}

/**
 * Run ONE asynchronous acquisition wave. `readFile` is the browser
 * filesystem adapter (OPFS, File System Access, or a worker message
 * bridge) supplied by the caller; it is awaited here and NOWHERE else —
 * commit and observe below are synchronous and never re-enter
 * asynchronous browser APIs.
 *
 * @param {(canonical: string) => Promise<string | null | undefined>} readFile
 * @param {string[]} canonicals
 * @returns {Promise<{files: Array<{canonical: string, content: string}>, missing: string[]}>}
 */
export async function acquireInputWave(readFile, canonicals) {
  if (typeof readFile !== "function") {
    throw new TypeError("acquireInputWave requires an async readFile adapter");
  }
  const files = [];
  const missing = [];
  for (const canonical of canonicals) {
    const content = await readFile(canonical);
    if (content == null) missing.push(canonical);
    else files.push({ canonical, content });
  }
  const failures = acquisitionWaveFailures(files, missing);
  if (failures.length > 0) {
    throw new Error(`incoherent acquisition wave: ${failures.join("; ")}`);
  }
  return { files, missing };
}

/**
 * Commit one acquired wave through the host's committed input-snapshot
 * route. The receipt binds the basis id, row counts, execution host
 * and engine version.
 *
 * @param {{commitInputSnapshot: (files: Array<{canonical: string, content: string}>, missing: string[]) => any}} host
 * @param {{files: Array<{canonical: string, content: string}>, missing: string[]}} wave
 */
export function commitInputSnapshot(host, wave) {
  const failures = acquisitionWaveFailures(wave.files, wave.missing);
  if (failures.length > 0) {
    throw new Error(`incoherent acquisition wave: ${failures.join("; ")}`);
  }
  return host.commitInputSnapshot(wave.files, wave.missing);
}

/**
 * Observe one canonical under a committed snapshot. A `needInputs`
 * status raises typed {@link NeedInputsError}; `absent` (the wave
 * probed the key and recorded it missing) and `file` return as-is —
 * distinct answers, never conflated with the acquisition demand.
 *
 * @param {{observeInputSnapshot: (basisId: string, canonical: string) => {status: string, content?: string}}} host
 * @param {string} basisId
 * @param {string} canonical
 */
export function observeInputSnapshot(host, basisId, canonical) {
  const observation = host.observeInputSnapshot(basisId, canonical);
  if (observation?.status === "needInputs") {
    throw new NeedInputsError([canonical]);
  }
  return observation;
}

/**
 * Release a committed snapshot the caller no longer observes. Every basis
 * stays addressable until released, so a host that commits repeated waves
 * releases the superseded basis after switching to the new one.
 *
 * @param {{releaseInputSnapshot: (basisId: string) => boolean}} host
 * @param {string} basisId
 * @returns {boolean} whether a snapshot was stored under `basisId`
 */
export function releaseInputSnapshot(host, basisId) {
  return host.releaseInputSnapshot(basisId) === true;
}

/** Canonicals for the next acquisition wave from a typed NeedInputs result. */
export function needInputsKeys(error) {
  if (error instanceof NeedInputsError) return [...error.keys];
  return [];
}

File: crates/verter_session/src/input_handoff.rs
//! Asynchronous input acquisition to committed-snapshot handoff.
//!
//! A browser host acquires file bytes asynchronously (OPFS, the File
//! System Access API, or a worker message) OUTSIDE any semantic
//! callback, then hands the acquired rows here in one synchronous call.
//! This seam commits ONE immutable [`InputBasis`] — positives plus
//! explicit negatives — through the canonical F1 constructor and binds
//! it behind a [`RequestInputBinding`] fence. Every later synchronous
//! semantic callback observes committed rows only.
//!
//! Two rules make the boundary honest:
//!
//! - **No acquisition capability lives here.** The handoff type holds
//!   no workspace handle, no loader, and no callback: a requested key
//!   that was never acquired surfaces as typed [`HandoffObserve::NeedInputs`]
//!   naming the missing canonical, never as a synchronous fetch attempt
//!   inside a resolver. The caller runs the NEXT asynchronous
//!   acquisition wave and commits the successor basis through the
//!   existing retry machinery.
//! - **Missing means probed-and-absent.** A key the acquisition wave
//!   probed and did not find is committed as a [`NegativeFact`] and
//!   observes as [`HandoffObserve::Absent`] — a complete-negative
//!   answer, distinct from `NeedInputs` the same way partial, pending
//!   and failed are distinct from complete-empty.

use std::sync::Arc;

use crate::input_basis::{InputBasis, LoadWave, NegativeFact, Observation, RequestInputBinding};

/// One asynchronously acquired file row handed to
/// [`CommittedInputHandoff::commit`].
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct AcquiredFile {
    /// Canonical identity of the acquired file.
    pub canonical: Arc<str>,
    /// File bytes captured by the asynchronous acquisition wave.
    pub content: Arc<str>,
}

/// One immutable committed handoff: the basis, its publication fence,
/// and nothing else.
#[derive(Debug, Clone)]
pub struct CommittedInputHandoff {
    binding: RequestInputBinding,
}

/// Typed observation outcome over a committed handoff.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum HandoffObserve {
    /// Committed file bytes.
    File {
        /// Canonical identity.
        canonical: Arc<str>,
        /// Captured content.
        content: Arc<str>,
    },
    /// The acquisition wave probed this key and recorded it absent.
    Absent {
        /// Canonical identity of the probed-missing input.
        canonical: Arc<str>,
    },
    /// The key was never acquired: the caller must run the next
    /// asynchronous acquisition wave. Not an error, not an absence,
    /// and never resolved by a read from here.
    NeedInputs {
        /// Canonical identity of the unacquired input.
        canonical: Arc<str>,
    },
}

impl HandoffObserve {
    /// The canonical this outcome answers for.
    #[must_use]
    pub fn canonical(&self) -> &str {
        match self {
            Self::File { canonical, .. }
            | Self::Absent { canonical }
            | Self::NeedInputs { canonical } => canonical,
        }
    }
}

impl CommittedInputHandoff {
    /// Commit one handoff from an acquisition wave's file rows and
    /// probed-missing keys.
    ///
    /// Duplicate identical rows collapse; a canonical both acquired and
    /// probed-missing, or acquired with two different contents, is
    /// refused closed by the canonical constructor.
    pub fn commit(
        files: impl IntoIterator<Item = AcquiredFile>,
        missing: impl IntoIterator<Item = Arc<str>>,
    ) -> Result<Self, crate::input_basis::CommitError> {
        let files: Vec<AcquiredFile> = files.into_iter().collect();
        let missing: Vec<Arc<str>> = missing.into_iter().collect();
        // The wave is the union of the rows themselves, so
        // `InputBasis::commit` sees one coherent wave covering every
        // positive and negative it is handed.
        let keys: Vec<Arc<str>> = files
            .iter()
            .map(|file| Arc::clone(&file.canonical))
            .chain(missing.iter().map(Arc::clone))
            .collect();
        let observations = files
            .into_iter()
            .map(|file| Observation::file(file.canonical, file.content));
        let negatives = missing.into_iter().map(NegativeFact::absent);
        let basis = InputBasis::commit(LoadWave::from_keys(keys), observations, negatives)?;
        Ok(Self {
            binding: RequestInputBinding::from_basis(basis),
        })
    }

    /// The bound request binding: the sole committed input and the
    /// fence that admits publication of this basis.
    #[must_use]
    pub fn binding(&self) -> &RequestInputBinding {
        &self.binding
    }

    /// The committed basis.
    #[must_use]
    pub fn basis(&self) -> &InputBasis {
        self.binding.basis()
    }

    /// Observe one key through the committed basis. Pure lookup: this
    /// type has no acquisition path, so an unrecorded key can only be
    /// the typed [`HandoffObserve::NeedInputs`] demand for the next
    /// asynchronous wave.
    #[must_use]
    pub fn observe(&self, canonical: &str) -> HandoffObserve {
        match self.binding.basis().observe(canonical) {
            Ok(observation) => match observation.kind() {
                crate::input_basis::ObservationKind::File { content } => HandoffObserve::File {
                    canonical: Arc::from(canonical),
                    content: Arc::clone(content),
                },
                crate::input_basis::ObservationKind::Directory { .. } => {
                    // File waves are the handoff's vocabulary; a
                    // directory row cannot arrive through
                    // `AcquiredFile`, so reaching here means a
                    // non-handoff basis was projected through this
                    // seam. Fail loudly rather than silently equating a
                    // directory listing with a missing input.
                    panic!("committed handoff projected a directory observation as a file wave")
                }
            },
            Err(crate::input_basis::ObserveError::Negative(_)) => HandoffObserve::Absent {
                canonical: Arc::from(canonical),
            },
            Err(crate::input_basis::ObserveError::Unrecorded) => HandoffObserve::NeedInputs {
                canonical: Arc::from(canonical),
            },
        }
    }

    /// Unrecorded keys among `keys`, sorted and deduplicated — the
    /// exact demand set for the next asynchronous acquisition wave.
    /// Already-committed positives and negatives are not re-demanded.
    #[must_use]
    pub fn next_acquisition_wave(&self, keys: impl IntoIterator<Item = Arc<str>>) -> LoadWave {
        self.binding.basis().discovery_wave(keys)
    }
}

#[cfg(test)]
#[path = "input_handoff_tests.rs"]
mod input_handoff_tests;
